### DISCUSSION TITLE
Ways to control server startup/bootstrap sequence in order to inject services?

### DISCUSSION THREAD

**Discussion by johanbrook** [remix-run/remix#2283] [Q&A] [answered]
Hi all!

I've scanned the docs and tried searching Issues/Discussions and google about this, but to no avail:

**I wonder if I can hook into the startup sequence of the Remix server, and create all my services in there.**

"Services" refers to databases, 3rd party APIs, micro services, etc.

Right now in my prototype, I've just done like the official guides do and initialise everything at top level:

```ts
// db.server.ts
export const db = createDb();
```
```ts
// some-route.tsx
import { db } from './db.server';

export const loader = () ⇒ {
   return db.users.findOne();
};
```

This is problematic because:

1) Unit testing becomes messy. Since a module might pull in something from the `db.server.ts` file, and the database gets initialised because it's on top level, etc.
2) Architecturally, it feels wrong to "hard code" stuff at top level. I just know I'm gonna need some kind of injection later on if the app grows and I want to do integration testing.

Would this be possible somehow in Remix? The server/client entrypoint files won't do, and the mysterious `server` config option in `remix.config.js` just makes Remix print

```
remix dev is not supported for custom servers
```

when running `npm run dev`.

Thank you!

**Comment by sergiodxa** [comment, marked as answer]
I think you can call your initializers inside entry.server and that should run before any other code.

Also if you use Express you control the server outside the Remix app and you could initialize them there.

**Reply by johanbrook** [reply]
I'll try calling them on top-level in `entry.server`, thanks!

Using Express would be nice for this specific case, for total control I guess. But perhaps not needed for small prototypes.
